### ISSUE TITLE
Precompiled SHA-1 Implementation

### ISSUE THREAD

**Issue by D-Nice** [ethereum/EIPs#180]
<pre>
  EIP: 180
  Title: Precompiled SHA-1 Implementation
  Author: Denis Milicevic &lt;denis@oraclize.it&gt;
  Status: Draft
  Type: Standards Track
  Created: 2016-11-24
</pre>

## Abstract

This draft EIP aims to introduce a new precompiled contract for the SHA-1 hashing algorithm alongside the launch of Metropolis. It is still a widely used algorithm for several protocols such as TLS, SSL and PGP to name a few along with its widespread use as a checksum. The implementation of SHA-1 as a precompile contract would provide the opportunity for on-chain verification of various services, namely oracles, at reasonable costs, as deploying and running it currently on the network is expensive for even basic inputs, and more important functions for verification such as HMAC-SHA1 are inconceivable for large inputs with the current and foreseeable gas limits and costs in place.
## Motivation

To provide the ability for on-chain verification of proofs such as TLSNotary, which the oracle service, Oraclize, makes use of to ensure oracled data was honestly fetched. The most intensive process for the verification of such proofs is from the HMAC-SHA1 data integrity check that occurs when decrypting the HTML within the proof. As it is raw HTML being passed through, one can expect certain instances to contain up to a few thousand bytes being passed as the input into the HMAC-SHA1. 

An optimized reference SHA-1 implementation in Solidity is available at the following [Gist link](https://gist.github.com/D-Nice/43d60515263f9854bd92e952dfc0dea5). 

Testing results on Ropsten, compiled with solc v0.4.7-nightly.2016.11.24+commit.851f8576

In terms of gas costs, it uses [287622 gas](https://testnet.etherscan.io/tx/0xf74debd5cf19b56343357094fc87797ac4ce0c0741e52135c9ae6f21ae784c6e) for an input up to 55 bytes in length and approximately an additional 255,000 per extra 64 bytes, increasing per block. Please refer to [WolframAlpha](http://www.wolframalpha.com/input/?i=1%2F2+(64430+%2B+510727+n+%2B+87+n%5E2)&lk=1&rawformassumption=%22ClashPrefs%22+-%3E+%7B%22Math%22%7D) for a link to the gas cost formula (applicable to inputs sized 1 byte below the current max block size). The gas formula of the contract is `1/2 (64430 + 510727 n + 87 n^2)` where n denotes the number of blocks, and block number can be derived from `n = ⌊( (L*8) + 1 + 64 )/512⌋ + 1`, where L is the byte length of the input.

As further examples, here are the gas costs for when the [second](https://testnet.etherscan.io/tx/0x2303e082445e97c77910cb8e9ae84117cc5df92b76ba05c61b3d1b153c2788d3) and [third](https://testnet.etherscan.io/tx/0x922a123ef517c117af835884a11549559934c1d1d8180471c02d2bc7b2543a51) blocks are used. If coupled as an HMAC, this would make on-chain verification very expensive for the smallest of inputs and not possible for larger ones. 
## Specification

If `block.number >= METROPOLIS_FORK_BLKNUM` 

Add a precompile at the next available precompile address, `0x0000...0005` or precompile address 5, which accepts a byte input of variable length but not exceeding  2^64 − 1 bits and returns a 20-byte (160-bit) message digest as the output. It should take the following form:
`SHA1(i ∈ B) ≡ o ∈ B20`
In terms of relative performance, when compared to its SHA-256 successor, the SHA-1 function has a comparable processing cost for an input up to 64 bytes. As the input size increases beyond this point, the SHA-1 function does get more noticeably efficient. [Benchmarks using OpenSSL](http://security.stackexchange.com/a/24331).

Thus far the gas costs for the precompiles appear to be safe as they have not been used in the DoS attacks. Therefore, the current SHA-256 costs would be a fair reference. The proposed gas cost for SHA-1 is based upon the SHA-256 formula in the Yellow Paper and defined as follows: 

```
ΞSHA1 ≡ ΞPRE where:
g = 62 + 8 * ⌊ |Id| / 32 ⌋
o[0..31] = SHA1(Id)
```

These are conservative gas costs, when compared to the overall better performance SHA-1 showcases in benchmarks. Here is the [spreadsheet](https://docs.google.com/spreadsheets/d/1tUPwyBnMJjqFnxMfoidF2yT3zxTqoPSefwFLZeFQsIg/edit?usp=sharing) which was used to help derive this formulation. The "Perf Target" column is based on the performance differences noted on the aforementioned OpenSSL benchmarks, and the gas cost was fine-tuned so that gas cost would approximate the performance curve, and very safe and simple values were chosen, with the BASE cost being 62 (2 higher than SHA-256), while the WORD cost is set to 8 (4 lower than SHA-256). 
## Rationale

The goal of this EIP is to be as conforming with the standards of the current protocol as possible therefore easing implementation between multiple clients as it would be a straightforward amendment to include an additional hashing function, and no changes to existing features occur. 
The gas costs were derived relatively from the current SHA-256 implementation, by examining benchmarks and then formulating a conservative  and safe gas cost for the SHA-1 precompile, as it is still not within the performance improvement target range of the used benchmarks. This should help ensure it does not become a DoS vector at some point, as with the current specification it is more expensive in terms of gas than its real computation costs.
Furthermore, the implementations by clients could be optimized by making use of SIMD instructions as specified by [Intel here](https://software.intel.com/en-us/articles/improving-the-performance-of-the-secure-hash-algorithm-1/).
There shouldn't be any backwards incompatibility, as no existing features are changed, and it should just be an additional feature. The only foreseeable incompatibility is if there is a current contract on Ethereum that uses the proposed address at `0x0000...0005` for something and checks if it is empty. If such an issue arises, the precompile could be tied to another address, although it would be optimal to follow the current precompile convention, in which case if the aforementioned contract that makes use of this address is not integral or not in use, it could be redeployed by its owner to use another arbitrary address for the check, as address 5 should not have any additional significance, aside from the fact that it is likely to be used for another precompile sometime in the future.

**Comment by D-Nice** [comment]
If applicable, and for disclosure purposes, I work for Oraclize.

**Comment by simonjanin** [comment]
This would be useful to me as well. +1

**Comment by tjayrush** [comment]
Is it wise it is to install a pre-compiled encryption algorithm that has already been shown to be exploitable: http://shattered.io

**Comment by MicahZoltu** [comment]
@D-Nice Can Oraclize it use a different hashing algorithm or are you stuck with SHA-1 due to the nature of how TLSNotary works with regards to TLS?  If TLSNotary of TLS web sites can't function without SHA1, then I think that is a compelling argument for putting a native SHA1 implementation into Ethereum.  If Oraclize it can do something to make it so SHA1 isn't necessary without sacrificing functionality, then I would vote to leave SHA1 out and have Oraclize it switch to a better algorithm.

**Comment by D-Nice** [comment]
@MicahZoltu We are indeed stuck with SHA-1, as TLSNotary works with TLS 1.0 and 1.1 whose protocol is dependent upon SHA-1. In TLS 1.2 a number of instances of SHA-1 are replaced with SHA-256, but there is still one part there as well, that's dependent on SHA-1. However, TLSNotary is only compatible with 1.0 and 1.1 at the moment, and likely won't work with TLS 1.2+ with the way it works.

**Comment by D-Nice** [comment]
To be clear, the potential precompile's utility isn't limited only to TLSNotary verification, but another issue I've just run into is SHA-1 RSA verification. People will be happy to see modexp be included in metropolis, but then find it useless for any certificates using SHA1withRSA.

**Comment by D-Nice** [comment]
To give an example of the impracticality of the contract for RSA verification, or most other for that matter; a  certificate with a body size of ~2kb would result in 33 data blocks and need ~8.5M gas for its SHA-1 hash to be computed.

**Comment by MapinguariJoulemeter** [comment]
I'd like to use asymmetric encryption in a smart contract as well as some other hash functions such as sha1 and md5 (not for use as a security measure). I support including the precompiled contracts as described by D-Nice.

**Comment by Arachnid** [comment]
As an alternative to a precompile, I've written an implementation in Solidity assembly: https://github.com/arachnid/solsha1

It requires roughly 56k gas per 512 bit block hashed.
